prolog/list_util.pl

```diff
@@ -129,20 +117,21 @@
 %  Xs = Take, Take = [] ;
 %  Xs = Take, Take = [_G3810].
 %  ==
-split_at(N,Xs,Take,Rest) :-
-    split_at_(Xs,N,Take,Rest).
+split_at(N, Xs, Take, Rest) :-
+    (  maplist(var, [N,Take])
+    -> split_at_(Xs, N, Take, Rest)
+    ;  once(split_at_(Xs, N, Take, Rest))
+    ).
 
-split_at_(Rest, 0, [], Rest) :- !. % optimization
+split_at_(Rest, 0, [], Rest).
 split_at_([], N, [], []) :-
-    % cannot optimize here because (+, -, -, -) would be wrong,
-    % which could possibly be a useful generator.
-    N > 0.
+    when(ground(N), N > 0).
+
 split_at_([X|Xs], N, [X|Take], Rest) :-
-    N > 0,
-    succ(N0, N),
+    when(ground(N), N > 0),
+    when(ground(N0), succ(N0, N)),
```

Comment:
Actually, I'm in favor of clpfd, I didn't go through with it because of my reservations about the extra dependency too. Regarding its advantages: If the library is to be worked on extensively in the future, perhaps this'll be advantageous to use clpfd anyway (development-wise). Moreover, this is one of the many intended use cases of clpfd anyway, to not use it now, in my opinion, is fatuous.